### ISSUE TITLE
standalone_2/io/secure_socket_renegotiate_test fails with RuntimeError

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#42771]
```
tools/test.py -n dartk-linux-release-x64 standalone_2/io/secure_socket_renegotiate_test
```

```
Client failed, stdout:

  stderr:
tests/standalone_2/io/secure_socket_renegotiate_client.dart:47:31: Error: The argument type 'Utf8Decoder' can't be assigned to the parameter type 'StreamTransformer<Uint8List, dynamic>'.
 - 'Utf8Decoder' is from 'dart:convert'.
 - 'StreamTransformer' is from 'dart:async'.
 - 'Uint8List' is from 'dart:typed_data'.
        socket.transform(utf8.decoder).transform(new LineSplitter()));
                              ^
```

There is a similar failure in `standalone/io/secure_socket_renegotiate_test` in NNBD weak mode.

If I change secure_socket_renegotiate_client.dart to use `utf8.decoder.bind(socket)` instead of `socket.transform(utf8.decoder)` (as CHANGELOG recommends), the test still fails.

```patch
--- a/tests/standalone_2/io/secure_socket_renegotiate_client.dart
+++ b/tests/standalone_2/io/secure_socket_renegotiate_client.dart
@@ -44,7 +44,7 @@ void runClient(int port) {
     expectEquals('CN=localhost', certificate.subject);
     expectEquals('CN=myauthority', certificate.issuer);
     StreamIterator<String> input = new StreamIterator(
-        socket.transform(utf8.decoder).transform(new LineSplitter()));
+        utf8.decoder.bind(socket).transform(new LineSplitter()));
     socket.writeln('first');
     input.moveNext().then((success) {
       expect(success);
```

```
Unhandled exception:
Expect.isTrue(false) fails.
#0      Expect._fail (package:expect/expect.dart:698:5)
#1      Expect.isTrue (package:expect/expect.dart:141:5)
#2      runServer.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///.../tests/standalone_2/io/secure_socket_renegotiate_test.dart:37:16)
```

/cc @tvolkert @lrhn @sortie @zichangg

**Comment by tvolkert** [comment]
Is there an underlying error in the [`StreamIterator`](https://github.com/dart-lang/sdk/blob/d0d263c4f10a480c15d1b02cf0032f612e388edb/tests/standalone_2/io/secure_socket_renegotiate_test.dart#L34) that's not showing up in the test logs when you make that change?  What happens if you change it to `socket.cast<List<int>>().transform(utf8.decoder)`?

**Comment by aam** [comment]
There are failing expectations on the client side of things(client is launched as a separate process) that cause this `Expect.isTrue(false)` to fail - those are easy to fix.

However there is bigger problem: implementation fo `SSLFilter::Renegotiate` seems to be doing nothing now since we migrated to boringssl here: https://github.com/dart-lang/sdk/blame/13bf8ff9f8d0f993673ac40bda08bde460ba71dd/runtime/bin/secure_socket.cc#L1009.
@whesse do you remember context for that decision? Was it just left to be implemented later? I expect we repeat handshake like we did before boringssl migration.

**Comment by whesse** [comment]
I'm not sure what the discussion was at that point, but renegotiation is now discontinued, and not a part of the current TLS standard.
Removing the ability to provide a named client certificate, I remember, because the use cases all included where the server specified what client certificate it wanted, and the client had it in its certificate database, so it didn't need to be passed separately.

But renegotiation to get a client certificate after the secure connection is already made is now discouraged (and not supported), according to all I can find on the Internet.  There seem to be fundamental insecurities with it, and most SSL client libraries (like Java) have removed it.

I think we should deprecate/remove the renegotiate method on our secure socket classes, and remove the tests.

**Comment by aam** [comment]
Given that it was broken for last five years, yeah probably nobody will notice if `renegotiate` method simply disappears.

One point though is that the boringssl which we build our api on does support renegotiation https://boringssl.googlesource.com/boringssl/+/65703b45c47bb71cfcb972edea00d82218cb0f9d/PORTING.md#tls-renegotiation - it does so with particular caveats which makes it safe to use. I thought that there might have been some reasons why didn't hook it up when we switched from openssl to boringssl.

**Comment by brianquinlan** [comment]
OK, I just discovered this issue independently. It looks like `renegotiate` [never did anything](https://github.com/dart-lang/sdk/blame/main/runtime/bin/secure_socket_filter.cc#L665)

I will deprecate it and update the docs.

**Comment by brianquinlan** [comment]
Fixed in cdcc221d6020f545b8a99dc3a8a8c14e3eb61949
